### ISSUE TITLE
RFC: Lean.Name.isUserName

### ISSUE THREAD

**Issue by joehendrix** [leanprover/lean4#2850]
### Proposal

There's a few places in Std and Mathlib where there is a desire to filter out auto-generated and other stable names (e.g., leanprover/std4#347 and leanprover/std4#355).

I think this notion could get upstreamed in the hopes that will ensure tactics use a consistent predicate.  The predicate I want is already in a [test case](https://github.com/leanprover/lean4/blob/b278172b7c8be73678e40cb7fc38c7b2db2bbe1b/tests/lean/run/printDecls.lean#L7-L11), but could get moved to `Lean/Data/Name.lean`:

Hear's the addition I would propose.  

```lean
/--
Return true if `declName` is a auxillary definition created by elaborator.
-/
def isAuxName : Name → Bool
| .str _ s => "match_".isPrefixOf s || "proof_".isPrefixOf s || "eq_".isPrefixOf s
| _ => false

/--
Return true if the name is a user facing name.

This returns false for names that are internal, numerical or appear auxillary created.
-/
def isUserName (declName : Name) : Bool :=
  declName.isStr
  && !declName.isAuxName
  && !declName.isInternal
```

### Community Feedback

I haven't done this, but it's hopefully just moving a definition.

### Impact

The potential impact is better consistency of ignoring definitions across tactics that want to show user facing names.

**Comment by Kha** [comment]
I think we should also double-check whether there is any reason to *not* mark these auxiliary names as internal, i.e. prefix them with an underscore

**Comment by nomeata** [comment]
Possibly Related: https://github.com/leanprover/lean4/issues/2719. Letrec'ed definitions probably should not be user visible by default, but sometimes they do

**Comment by joehendrix** [comment]
I'll mention @leodemoura to see if he can answer the auxiliary question.

To summarize, I'd like to introduce a name predicate `isUserName` that all tactics/commands can use to distinguish names users should ordinarily see (and be able to reference without them changing unexpectedly).  The definition is pulled from a test file.  This is useful for linters and search commands and tactics like (`#print prefix`).

Some questions I have are:
* Should auxiliary names like `Name.match_1` be made internal (e.g., `Name._match_1`?
* Should `isUserName` return false if there is a `.num` component in any part of the path?  Right now the logic only checks that the name ends with a `.str` component.
* Are there other patterns `isUserName` should fail on?

**Comment by digama0** [comment]
One case where `.num` components appear in things that are not necessarily internal is in definitions created by macros and private definitions.

I still think we should have some kind of explicit `@[automatically_generated]` attribute (this is what rust does BTW) to standardize this rather than using underscores, because sometimes autogenerated definitions are also intended for use by users.

**Comment by semorrison** [comment]
For comparison, Mathlib currently has 3 separate(!) `Name.isBlacklisted` functions:

* `Mathlib.Tactic.Find`:
```
private def isBlackListed (declName : Name) : MetaM Bool := do
  let env ← getEnv
  pure $ declName.isInternal
   || isAuxRecursor env declName
   || isNoConfusion env declName
  <||> isRec declName
  <||> isMatcher declName
```

* `Mathlib.Lean.Expr.Basic`
```
def isBlackListed {m} [Monad m] [MonadEnv m] (declName : Name) : m Bool := do
  if declName == ``sorryAx then return true
  if declName matches .str _ "inj" then return true
  if declName matches .str _ "noConfusionType" then return true
  let env ← getEnv
  pure $ declName.isInternal'
   || isAuxRecursor env declName
   || isNoConfusion env declName
  <||> isRec declName <||> isMatcher declName
```

* `Mathlib.Lean.Name`
```
private def isBlackListed (declName : Name) : CoreM Bool := do
  if declName.toString.startsWith "Lean" then return true
  let env ← getEnv
  pure $ declName.isInternal'
   || isAuxRecursor env declName
   || isNoConfusion env declName
  <||> isRec declName <||> isMatcher declName
```

Along with
```
def isInternal' (declName : Name) : Bool :=
  declName.isInternal ||
  match declName with
  | .str _ s => "match_".isPrefixOf s || "proof_".isPrefixOf s
  | _        => true
```

What a mess!

**Comment by digama0** [comment]
Note: just because there are many functions doesn't necessarily imply that they should be unified into just one. Quite often the function was written because an upstream function exists but doesn't *quite* do what is needed for some particular use case. For example the one from `Mathlib.Lean.Expr.Basic` counts `sorry` as "blacklisted" but of course it is not "internal". My guess is that this is being used from some proof search tactic like `exact?` which wants to blacklist certain theorems, and this is not the same usecase as e.g. showing in doc-gen.

**Comment by semorrison** [comment]
Yes. Presumable we will still need several, but it would be good if they were more compositional!

**Comment by joehendrix** [comment]
@digama0 Do you have pointers to code where a non-internal `.num` component is being synthesized by a tactic or macro?

I was initially thinking tuples, but those seem to desugar into `Prod`.  I think perhaps the tactics should be changed to use a more stable name.

**Comment by digama0** [comment]
Definitions made in macros are hygienic and hence contain `num` components:
```lean
import Lean

macro "def_stuff" : command => `(command| 
  def foo := 1
)

set_option trace.Elab.definition true in
def_stuff

#print foo

open Lean
macro "print_foo" : command => do
  let id := Name.appendCore (.str `foo "_@") (.num `Test._hyg 225)
  `(command| #print $(mkIdent id))

print_foo
```
The most common place you see these definitions are in `deriving` instances:
```lean
import Lean

set_option trace.Elab.definition true in
structure Foo where
  a : Nat
  deriving Lean.FromJson

#synth Lean.FromJson Foo -- instFromJsonFoo
#print instFromJsonFoo -- { fromJson? := fromJsonFoo✝ }

open Lean
macro "print_from_json_foo" : command => do
  let id := Name.appendCore 
    (.str (.num `_private.Test 0 ++ `fromJsonFoo) "_@") (.num `Test._hyg 18)
  `(command| #print $(mkIdent id))

print_from_json_foo
-- private def fromJsonFoo._@.Test._hyg.18 : Json → Except String Foo :=
-- fun j => do
--   let a ← Except.mapError (fun s => toString `Foo ++ "." ++ toString `a ++ ": " ++ s) (Json.getObjValAs? j Nat "a")
--   pure { a := a }
```
Of course these are probably intentionally internal (although this does make proving properties or unfolding these definitions really annoying).
